frontend/src/utils/api.ts

```diff
@@ -436,7 +456,18 @@ function createAPIError(response) {
   return error;
 }
 
-function attachData(response) {
+interface FMSResponse<T = any> extends Response {
+  isJSON: boolean
+  data: T
+  filename?: string
+}
+interface JsonResponse<T = any> extends FMSResponse<T> { isJSON: true }
+interface ArrayBufferResponse extends FMSResponse<ArrayBuffer> { isJSON: false }
+interface StringResponse extends FMSResponse<string> { isJSON: false }
+interface EmptyResponse extends FMSResponse<Record<string, never>> { isJSON: false }
+type ResponseWithData<T = any> = JsonResponse<T> | ArrayBufferResponse | StringResponse | EmptyResponse
```

Comment:
Not sure i see the point of a type of responsewithdata that includes empty responses ... Maybe call it AnythingGoesResponse ?

Reply:
It's an anomylous response because that kind of response is only returned when reading from the response body fails. I guess I could call it AttachDataErrorResponse 🤷

Reply:
So far I have never seen that kind of response, and most frontend codes assume successful body read.

Reply:
Hmm... maybe I should throw an exception instead of a body every fails to be read or use `Promise.reject`.